src/services/refact/docker.ts

```diff
@@ -0,0 +1,301 @@
+        return { data: response.data };
+      },
+    }),
+  }),
+});
+
+export type DockerActionResponse = {
+  success: boolean;
+  output: string;
+};
+
+/**
+ * Represents the payload for a Docker action endpoints.
+ * @param action Docker action for a specific operation (start, stop, kill, remove)
+ * @param container This can be either the container name or the container ID.
+ */
+export type DockerActionPayload = {
+  action: "start" | "stop" | "kill" | "remove";
+  container: string;
+};
+
+function isDockerActionResponse(json: unknown): json is DockerActionResponse {
+  if (!json) return false;
+  if (typeof json !== "object") return false;
+  if (!("success" in json)) return false;
+  if (typeof json.success !== "boolean") return false;
+  if (!("output" in json)) return false;
+  if (typeof json.output !== "string") return false;
+  return true;
+}
+
+type DockerRequestBody = {
+  label?: string;
+  image?: string;
+};
+
+export type DockerContainer = {
+  id: string;
+  name: string;
+  status: string;
+  created: string;
+  user: string;
+  env: string[];
+  command: string[];
+  image: string;
+  working_dir: string;
+  labels: DockerLabels;
+  ports: DockerPorts;
+};
+
+type DockerLabels = NonNullable<unknown>;
+
+// TODO: make types for ports
+type DockerPorts = NonNullable<unknown>;
+
+// TODO: make type guards better
+type DockerContainersResponse = {
+  containers: DockerContainer[];
+};
+
+function isDockerContainersResponse(
+  json: unknown,
+): json is DockerContainersResponse {
+  if (
+    !json ||
+    typeof json !== "object" ||
+    !Array.isArray((json as DockerContainersResponse).containers)
+  ) {
+    return false;
+  }
+  const containers = (json as DockerContainersResponse).containers;
+  return containers.every(isDockerContainer);
+}
+
+function isDockerContainer(json: unknown): json is DockerContainer {
+  if (!json || typeof json !== "object") return false;
+
+  const container = json as DockerContainer;
+
+  if (typeof container.id !== "string") return false;
+  if (typeof container.name !== "string") return false;
+  if (typeof container.status !== "string") return false;
+  if (typeof container.created !== "string") return false;
+  if (typeof container.user !== "string") return false;
+  if (
+    !Array.isArray(container.env) ||
+    !container.env.every((e) => typeof e === "string")
+  )
+    return false;
+
+  if (
+    !Array.isArray(container.command) ||
+    !container.command.every((c) => typeof c === "string")
+  )
+    return false;
+
+  if (typeof container.image !== "string") return false;
+  if (typeof container.working_dir !== "string") return false;
+  if (!isDockerLabels(container.labels)) return false;
+  if (!isDockerPorts(container.ports)) return false;
+  return true;
+}
+
+function isDockerLabels(json: unknown): json is DockerLabels {
+  // Since DockerPorts is defined as NonNullable<unknown>, we don't have specific structure to validate. Just checking, that it's not null | undefined
```

Comment:
Are these unknown or likely to change?

Reply:
To be honest, we don't even render them, the only requirement for `labels` and `ports` is to be an object (`NonNullable<unknown>`)